### PR TITLE
203 - Add ability to upload images

### LINKED ISSUE #203: Add ability to display images for assets
# Acceptance Criteria

Given when a user is viewing asset details
Then the user should be able to see the image in grid component form.

1. This needs some work to be done to connect with s3 bucket storage so that users can use the bucket storage ability to view / add / edit images. Currently, this is not in the works. We would actually prioritize this at a later date and revisit the requirement.

### PR DESCRIPTION
**Pull request by mohit2530**
Update docker compose file

- updated docker compose file to support housing of local_images folder in the application. this will allow us to keep uploaded images separate that the application itself.

Added filePath directory in main api.go layer

- the go service will serve up the file server api. this will support loading and uploading images to and from the local_uploads folder.

Added support to read images from volume

- added go api to retrieve files from local_uploads folder
- added local_uploads folder to gitignore so that we don't mistakely upload images
- added two api endpoints to return images in the filesystem. this part will require more cleanup. currently files are stored within local_uploads/userID. if the user id changes we also need to be able to update that, and we also need to be able to know / limit the size of the uploads. Furthermore, when the container is closed, we need to make sure images are removed so the space is saved and this also needs to work in other docker env files. currently support is built for local dev setup. later on, we will need the same support in prod env.

Add permission script and add new apis

- added new permission script to allow read / write access to local_uploads folder. since docker uses root as the main user, we need to add the proper permissions so it can upload data into the local_uploads folder.

Update existing create api for file system

- altered api to create the new file inside the local_uploads folder.

Build frontend for image picker

- added ability to upload files and view them from the backend
- added saga and redux actions for image
- added support with formData for images.
- updated api layer to support association handler for images
- updated integration for inventories

Create sql file to support image url

- added new sql script to support images upload
- added image and display in inventories

Address displaying images for inventories

- updated api call in GridComponent to support only if there is an associated item url.
- added takeEvery for redux action so that we don't leave behind the selected images for certain
components.
- added 2mb limit to image upload
- added check to ensure the file is an image type.
- added snackbar to failure messages.

Tech debt

- tech debt cleanup
- added redirect to home after error has occurred
- added new swagger changes

# Screenshots / Visuals

image
image
image

Closes #203

### CODE REVIEW

File: apilayer/handler/FileServerHandler.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+	directoryPath := filepath.Join("..", "local_uploads", userID)
+	if _, err := os.Stat(directoryPath); os.IsNotExist(err) {
+		log.Printf("Unable to view associated files. error: %+v", err)
+		rw.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(rw).Encode(nil)
+		return
+	}
+
+	files, err := os.ReadDir(directoryPath)
```
[the remark this answers is not in the document]
Comment (mohit2530) [reply apilayer/handler/FileServerHandler.go:52]: Path is not created by user. path should already exists. User values does not affect path.

File: apilayer/handler/FileServerHandler.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		log.Printf("Unable to retrieve selected file without filename")
+		rw.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(rw).Encode(nil)
+		return
+	}
+
+	filePath := filepath.Join("..", "local_uploads", userID, filename)
+	log.Printf("Serving file from path: %s", filePath)
+
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
```
[the remark this answers is not in the document]
Comment (mohit2530) [reply apilayer/handler/FileServerHandler.go:110]: Path is not created by user. path should already exists. User values does not affect path.

File: apilayer/service/FileUpload.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+const Upload_Directory = "local_uploads"
+
+// UploadAndAssociateFile ...
+//
+// upload image to the local system and associate it with the passed in id. the workflow can be associated with various tables we also pass in the type of association.
+func UploadAndAssociateFile(fileToProcess multipart.File, itemType rune, userID string, itemID string, user string) (string, error) {
+	uploadDir := filepath.Join("..", Upload_Directory, userID)
+	if err := os.MkdirAll(uploadDir, os.ModePerm); err != nil {
```
[the remark this answers is not in the document]
Comment (mohit2530) [reply apilayer/service/FileUpload.go:23]: Path is not created by user. path should already exists. User values does not affect path.

File: apilayer/service/FileUpload.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		log.Printf("unable to create user directory for images. erorr: %+v", err)
+		return "", err
+	}
+
+	fileID := uuid.New().String()
+	dstFilePath := filepath.Join(uploadDir, fmt.Sprintf("%s.png", fileID))
+	dstFile, err := os.Create(dstFilePath)
```
[the remark this answers is not in the document]
Comment (mohit2530) [reply apilayer/service/FileUpload.go:30]: Path is not created by user. path should already exists. User values does not affect path.

File: apilayer/handler/FileServerHandler.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+	if len(userID) <= 0 {
+		log.Printf("Unable to view files associated with the selected user.")
+		rw.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(rw).Encode(nil)
+		return
+	}
+
+	directoryPath := filepath.Join("..", "local_uploads", userID)
+	if _, err := os.Stat(directoryPath); os.IsNotExist(err) {
```
[the remark this answers is not in the document]
Comment (mohit2530) [reply apilayer/handler/FileServerHandler.go:45]: not applicable. user does not modify or alter paths.

**Review by earmuff-jam — APPROVED** [review]

**Pull request merged**